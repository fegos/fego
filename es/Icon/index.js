'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

exports.default = Icon;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

require('./font.js');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectWithoutProperties(obj, keys) { var target = {}; for (var i in obj) { if (keys.indexOf(i) >= 0) continue; if (!Object.prototype.hasOwnProperty.call(obj, i)) continue; target[i] = obj[i]; } return target; } /**
                                                                                                                                                                                                                              * 图标组件
                                                                                                                                                                                                                              * @author esky
                                                                                                                                                                                                                              */


function Icon(props) {
  var name = props.name,
      _props$className = props.className,
      className = _props$className === undefined ? '' : _props$className,
      spin = props.spin,
      size = props.size,
      width = props.width,
      height = props.height,
      _props$hrefPrefix = props.hrefPrefix,
      hrefPrefix = _props$hrefPrefix === undefined ? 'nsicon' : _props$hrefPrefix,
      style = props.style,
      rest = _objectWithoutProperties(props, ['name', 'className', 'spin', 'size', 'width', 'height', 'hrefPrefix', 'style']);

  var cls = (className + ' nsicon ' + (spin || name === 'loading' ? 'nsicon-spin' : '')).trim();
  var _style = _extends({}, style, { width: size || width, height: size || height });
  /**
  <!-- REACT JSX: -->
  <svg> <use xlinkHref='/svg/svg-sprite#my-icon' /> </svg>
  <!-- RENDERS AS: -->
  <svg> <use xlink:href="/svg/svg-sprite#my-icon"></use> </svg>
    */
  return _react2.default.createElement(
    'svg',
    _extends({ className: cls, style: _style, 'aria-hidden': 'true' }, rest),
    _react2.default.createElement('use', { xlinkHref: '#' + hrefPrefix + '-' + name })
  );
}
/**
symbol引用

这是一种全新的使用方式，应该说这才是未来的主流，也是平台目前推荐的用法。相关介绍可以参考这篇文章 这种用法其实是做了一个svg的集合，与另外两种相比具有如下特点：

支持多色图标了，不再受单色限制。
通过一些技巧，支持像字体那样，通过font-size,color来调整样式。
兼容性较差，支持 ie9+,及现代浏览器。
浏览器渲染svg的性能一般，还不如png。
 */