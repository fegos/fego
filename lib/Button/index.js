'use strict';

exports.__esModule = true;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _Icon = require('../Icon');

var _Icon2 = _interopRequireDefault(_Icon);

var _PropsType = require('./PropsType');

var _util = require('./util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var Button = function (_React$Component) {
  _inherits(Button, _React$Component);

  function Button(props) {
    _classCallCheck(this, Button);

    var _this = _possibleConstructorReturn(this, _React$Component.call(this, props));

    _this.handleClick = function (e) {
      if (_this.state.clicked) return;
      _this.setState({ clicked: true });
      clearTimeout(_this.timeout);
      _this.timeout = setTimeout(function () {
        return _this.setState({ clicked: false });
      }, 300);
      var onClick = _this.props.onClick;

      if (onClick) {
        onClick(e);
      }
    };

    _this.state = {
      clicked: false,
      loading: props.loading
    };
    return _this;
  }

  Button.prototype.componentWillReceiveProps = function componentWillReceiveProps(nextProps) {
    var _this2 = this;

    var currentLoading = this.props.loading;
    var loading = nextProps.loading;

    if (currentLoading) {
      clearTimeout(this.delayTimeout);
    }
    if (typeof loading !== 'boolean' && loading && loading.delay) {
      this.delayTimeout = setTimeout(function () {
        return _this2.setState({ loading: loading });
      }, loading.delay);
    } else {
      this.setState({ loading: loading });
    }
  };

  Button.prototype.componentWillUnmount = function componentWillUnmount() {
    this.timeout && clearTimeout(this.timeout);
    this.delayTimeout && clearTimeout(this.delayTimeout);
  };

  Button.prototype.render = function render() {
    var _classNames;

    var _props = this.props,
        type = _props.type,
        shape = _props.shape,
        _props$size = _props.size,
        size = _props$size === undefined ? '' : _props$size,
        className = _props.className,
        htmlType = _props.htmlType,
        children = _props.children,
        icon = _props.icon,
        prefixCls = _props.prefixCls,
        disabled = _props.disabled,
        dashed = _props.dashed,
        style = _props.style;
    var _state = this.state,
        loading = _state.loading,
        clicked = _state.clicked;

    var classes = (0, _classnames2.default)(prefixCls, className, (_classNames = {}, _classNames[prefixCls + '-' + type] = type, _classNames[prefixCls + '-' + shape] = shape, _classNames[prefixCls + '-lg'] = size === 'large', _classNames[prefixCls + '-sm'] = size === 'small', _classNames[prefixCls + '-dashed'] = dashed, _classNames[prefixCls + '-icon-only'] = !children && icon && !loading, _classNames[prefixCls + '-loading'] = loading, _classNames[prefixCls + '-clicked'] = clicked, _classNames));
    var iconName = loading ? 'loading' : icon;
    var iconEl = iconName ? _react2.default.createElement(_Icon2.default, { name: iconName }) : null;
    var needInserted = _react2.default.Children.count(children) === 1 && (!iconName || iconName === 'loading');
    var kids = _react2.default.Children.map(children, function (child) {
      return (0, _util2.default)(child, needInserted);
    });
    return _react2.default.createElement(
      'button',
      { type: htmlType, className: classes, onClick: this.handleClick, disabled: disabled, style: style },
      iconEl,
      kids
    );
  };

  return Button;
}(_react2.default.Component);

exports.default = Button;


Button.defaultProps = _PropsType.defaultProps;
Button.propTypes = _PropsType.propTypes;