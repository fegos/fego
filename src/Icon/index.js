/**
 * 图标组件
 * @author esky
 */
import React from 'react'
import './font'
export default function Icon(props) {
  const { name, className = '', spin, size, width, height, hrefPrefix = 'nsicon', style, ...rest } = props;
  const cls = `${className} nsicon ${spin || name === 'loading' ? 'nsicon-spin' : ''}`.trim();
  let _style = { ...style, width: size || width, height: size || height };
  /**
  <!-- REACT JSX: -->
  <svg> <use xlinkHref='/svg/svg-sprite#my-icon' /> </svg>
  <!-- RENDERS AS: -->
  <svg> <use xlink:href="/svg/svg-sprite#my-icon"></use> </svg>
    */
  return <svg className={cls} style={_style} aria-hidden="true" {...rest}><use xlinkHref={`#${hrefPrefix}-${name}`} /></svg>
}
/**
symbol引用

这是一种全新的使用方式，应该说这才是未来的主流，也是平台目前推荐的用法。相关介绍可以参考这篇文章 这种用法其实是做了一个svg的集合，与另外两种相比具有如下特点：

支持多色图标了，不再受单色限制。
通过一些技巧，支持像字体那样，通过font-size,color来调整样式。
兼容性较差，支持 ie9+,及现代浏览器。
浏览器渲染svg的性能一般，还不如png。
 */
