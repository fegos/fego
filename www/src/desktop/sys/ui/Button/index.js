import React, { Component } from 'react';
import { Button } from 'fego';
import style from './index.less';

export default class Page extends Component {
  state = {}
  render() {
    return (
      <div className={style.page}>
        <div className="section">
          <h3>默认按钮：</h3>
          <Button type="default">default Button</Button>
          <Button disabled>disabled default Button</Button>
        </div>
        <div className="section">
          <h3>加载中状态：</h3>
          <Button loading>loading Button</Button>
        </div>
        <div className="section">
          <h3>主按钮：</h3>
          <Button type="primary">primary Button</Button>
          <Button type="primary" disabled>disabled primary Button</Button>
        </div>
        <div className="section">
          <h3>危险按钮：</h3>
          <Button type="danger">danger Button</Button>
          <Button type="danger" disabled>disabled danger Button</Button>
        </div>
        <div className="section ghost">
          <h3>幽灵按钮：</h3>
          <Button type="ghost">ghost Button</Button>
          <Button type="ghost" disabled>disabled ghost Button</Button>
        </div>
        <div className="section">
          <h3>虚线按钮：</h3>
          <Button dashed>dashed Button</Button>
        </div>
        <div className="section">
          <h3>按钮尺寸：</h3>
          <Button size="small">small Button</Button>
          <Button>default size Button</Button>
          <Button size="large">large Button</Button>
        </div>
        <div className="section">
          <h3>圆形按钮：</h3>
          <Button shape="circle" icon="plus" />
          <Button shape="circle" dashed>btn</Button>
        </div>
      </div>
    );
  }
}
