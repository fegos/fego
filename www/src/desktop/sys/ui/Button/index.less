.page {
	padding: 20px 10px;
	padding-left: 0;
	:global {
		.section {
			padding: 10px;
			margin-top: 10px;
			&.ghost {
				background: #dbdbdb;
			}
			& > span {
				display: inline-block;
				width: 100px;
			}
		}
		.ns-btn {
			margin: 0 20px;
		}
	}
}