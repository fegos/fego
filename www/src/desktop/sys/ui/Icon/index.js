import React, { Component } from 'react';
import { Icon } from 'fego';
import style from './index.less';

const icons = [
  'question-fill',
  'error',
  'info',
  'warning',
  'success',
  'home',
  'minus',
  'plus',
  'yes',
  'no',
];

export default class Page extends Component {
  state = {}
  render() {
    return (
      <div className={style.page}>
        <div className="section f-cb">
          {icons.map(icon => (
            <div className="item" key={icon}>
              <Icon name={icon} />
              <span className={style.name}>{icon}</span>
            </div>
          ))}
        </div>
      </div>
    );
  }
}
