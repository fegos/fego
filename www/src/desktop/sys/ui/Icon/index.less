.page {
  :global {
    .item {
      width: 100px;
      height: 100px;
      float: left;
      margin-right: 20px;
      display: flex;
      flex-direction: column;
      justify-content: center;
      align-items: center;
      span {
        margin-top: 15px;
      }
    }
  }
}
