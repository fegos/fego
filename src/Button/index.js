import React from 'react'
import classNames from 'classnames'
import Icon from '../Icon'
import { defaultProps, propTypes } from './PropsType'
import insertSpace from './util'

export default class Button extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			clicked: false,
			loading: props.loading,
		}
	}
	handleClick = e => {
		if (this.state.clicked) return;
		this.setState({ clicked: true });
		clearTimeout(this.timeout);
		this.timeout = setTimeout(() => this.setState({ clicked: false }), 300);
		const onClick = this.props.onClick;
		if (onClick) {
			onClick(e);
		}
	}
	componentWillReceiveProps(nextProps) {
		const currentLoading = this.props.loading;
		const loading = nextProps.loading;
		if (currentLoading) {
			clearTimeout(this.delayTimeout);
		}
		if (typeof loading !== 'boolean' && loading && loading.delay) {
			this.delayTimeout = setTimeout(() => this.setState({ loading }), loading.delay);
		}
		else {
			this.setState({ loading });
		}
	}
	componentWillUnmount() {
		this.timeout && clearTimeout(this.timeout);
		this.delayTimeout && clearTimeout(this.delayTimeout);
	}
	render() {
		const { type, shape, size = '', className, htmlType, children, icon, prefixCls, disabled, dashed, style } = this.props;
		const { loading, clicked } = this.state;
		const classes = classNames(prefixCls, className, {
			[`${prefixCls}-${type}`]: type,
			[`${prefixCls}-${shape}`]: shape,
			[`${prefixCls}-lg`]: size === 'large',
			[`${prefixCls}-sm`]: size === 'small',
			[`${prefixCls}-dashed`]: dashed,
			[`${prefixCls}-icon-only`]: !children && icon && !loading,
			[`${prefixCls}-loading`]: loading,
			[`${prefixCls}-clicked`]: clicked,
		});
		const iconName = loading ? 'loading' : icon;
		const iconEl = iconName ? <Icon name={iconName} /> : null;
		const needInserted = React.Children.count(children) === 1 && (!iconName || iconName === 'loading');
		const kids = React.Children.map(children, child => insertSpace(child, needInserted));
		return (<button type={htmlType} className={classes} onClick={this.handleClick} disabled={disabled} style={style} >
			{iconEl}{kids}
		</button>);
	}
}

Button.defaultProps = defaultProps;
Button.propTypes = propTypes;