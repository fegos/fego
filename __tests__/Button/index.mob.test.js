import React from 'react';
import Renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import Button from 'Button/index.mob';
import Icon from 'Icon';

describe('button test', () => {
  it('button snapshot', () => {
    const button = Renderer.create((<Button>button</Button>)).toJSON();
    expect(button).toMatchSnapshot();
  });

  it('different type of button', () => {
    const defaultButton = shallow(<Button type="default">button</Button>);
    const primaryButton = shallow(<Button type="primary">button</Button>);
    const dangerButton = shallow(<Button type="danger">button</Button>);
    const ghostButton = shallow(<Button type="ghost">button</Button>);

    expect(defaultButton.find('.ns-btn-default')).toHaveLength(1);
    expect(defaultButton.find('.ns-btn-primary')).toHaveLength(0);
    expect(primaryButton.find('.ns-btn-primary')).toHaveLength(1);
    expect(primaryButton.find('.ns-btn-danger')).toHaveLength(0);
    expect(dangerButton.find('.ns-btn-danger')).toHaveLength(1);
    expect(dangerButton.find('.ns-btn-ghost')).toHaveLength(0);
    expect(ghostButton.find('.ns-btn-ghost')).toHaveLength(1);
    expect(ghostButton.find('.ns-btn-default')).toHaveLength(0);
  });

  it('button with icon', () => {
    const button = mount(<Button icon="home">button</Button>);
    expect(button.find('.nsicon')).toHaveLength(1);
  });

  it('loading button', () => {
    const button = shallow(<Button loading>button</Button>);
    expect(button.find(Icon)).toHaveLength(1);
    expect(button.find('.ns-btn-loading')).toHaveLength(1);
  });

  it('fire touchstart event', () => {
    const onClickFunc = jest.fn();
    const button = mount(<Button onClick={onClickFunc}>button</Button>);
    const disabledButton = mount(<Button onClick={onClickFunc} disabled>button</Button>);
    expect(button.state('clicked')).toBe(false);
    expect(disabledButton.state('clicked')).toBe(false);
    button.simulate('touchstart');
    button.simulate('touchstart');
    disabledButton.simulate('touchstart');
    // 移动端的 touchstart 事件，但是 enzyme 貌似不能区分，所以此处的 touchstart 并不是手指按下就不松开了
    // 而是按下-松开这整个过程都有，只不过只会显示的调用 onTouchStart 的回调罢了
    expect(button.state('clicked')).toBe(true);
    expect(disabledButton.state('clicked')).toBe(false);
    expect(onClickFunc).toHaveBeenCalledTimes(0);
  });

  it('fire touchend event', () => {
    const onClickFunc = jest.fn();
    const button = mount(<Button onClick={onClickFunc}>button</Button>);

    button.simulate('touchend');

    expect(onClickFunc).toHaveBeenCalledTimes(1);
  });

  it('fire willReceiveProps lifecycle', () => {
    const button = mount(<Button loading>button</Button>);
    expect(button.state('loading')).toBe(true);
    button.setProps({
      loading: false,
    });
    expect(button.state('loading')).toBe(false);
  });
});
