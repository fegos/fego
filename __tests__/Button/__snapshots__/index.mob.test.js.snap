// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button test button snapshot 1`] = `
<button
  className="ns-btn ns-btn-default"
  disabled={false}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
  style={undefined}
  type="button"
>
  <span>
    button
  </span>
</button>
`;
